File: core/concurrency.py
"""Bounded-concurrency worker pool for parallel API fetches."""

from concurrent.futures import ThreadPoolExecutor, as_completed


def map_pool(items, concurrency, worker_fn):
    if not items:
        return []

    limit = max(1, min(concurrency, len(items)))
    results = []

    with ThreadPoolExecutor(max_workers=limit) as executor:
        futures = {executor.submit(worker_fn, item): item for item in items}
        for future in as_completed(futures):
            try:
                value = future.result()
                if value is not None:
                    results.append(value)
            except Exception:
                pass

    return results

File: core/concurrency.js
/**
 * Bounded-concurrency async pool for parallel API fetches.
 */

async function mapPool(items, concurrency, workerFn) {
  if (!items || items.length === 0) return [];

  const limit = Math.max(1, Math.min(concurrency, items.length));
  const results = new Array(items.length);
  let nextIndex = 0;

  async function runWorker() {
    while (nextIndex < items.length) {
      const current = nextIndex++;
      try {
        results[current] = await workerFn(items[current], current);
      } catch (err) {
        results[current] = null;
      }
    }
  }

  await Promise.all(Array.from({ length: limit }, () => runWorker()));
  return results.filter((r) => r != null);
}

module.exports = { mapPool };